rfc-087-dealing-with-errors.md

```diff
@@ -0,0 +1,105 @@
+
+This describes how we treat errors on GOV.UK
+
+## Problem
+
+We've recently migrated to a new error tracking service, Sentry. This provides us an opportunity to rethink how we treat errors.
+
+## Proposal
+
+There are 2 principles:
+
+### 1. When something goes wrong, we should be notified
+
+Applications should report exceptions to Sentry. Applications must not swallow errors.
+
+### 2. Applications should not error
+
+The goal of GOV.UK is that applications should not error. When something goes wrong it should be fixed.
+
+## Classifying errors
+
+### Bug
+
+A code change makes the application crash.
+
+Desired behaviour: error is sent to Sentry, developers are notified and fix the error. Developers mark the error in Sentry as `Resolved`. This means a recurrence of the error will alert developers again.
+
+### Intermittent errors without user impact
+
+Frontend applications often see timeouts when talking to the content-store.
+
+There's no or little user impact because the request will be answered by the caching layer.
+
+Example: <https://sentry.io/govuk/app-finder-frontend/issues/352985400>
+
+Desired behaviour: error is not sent to Sentry. Instead, we rely on Smokey and Icinga checks to make sure we the site functions.
+
+### Intermittent errors with user impact
+
+Publishing applications sometimes see timeouts when talking to publishing-api. This results in the publisher seeing an error page and possibly losing data.
+
+Example: <https://sentry.io/govuk/app-content-tagger/issues/367277928>
+
+Desired behaviour: TBD
+
+### Intermittent retryable errors
+
+Sidekiq worker sends something to the publishing-api, which times out. Sidekiq retries, the next time it works.
+
+Desired behaviour: errors are not reported to Sentry until retries are exhausted. See [this PR for an example](https://github.com/alphagov/content-performance-manager/pull/353).
+
+Relevant: https://github.com/getsentry/raven-ruby/pull/784
+
+### Expected environment-based errors
+
+MySQL errors on staging while data sync happens.
+
+Example: <https://sentry.io/govuk/app-whitehall/issues/343619055>
+
+Desired behaviour: our environment is set up such that these errors do not occur.
+
+### Bad request errors
+
+User makes a request the application can't handle ([example][bad-request]).
+
+Often happens in [security checks](https://sentry.io/govuk/app-frontend/issues/400074979).
+
+Example: <https://sentry.io/govuk/app-frontend/issues/400074979>
+
+Desired behaviour: user gets feedback, error is not reported to Sentry
+
+[bad-request]: https://sentry.io/govuk/app-service-manual-frontend/issues/400074003
+
+### Incorrect bubbling up of errors
+
+Rummager crashes on date parsing, returns 500, which is passed on directly in finder-frontend.
+
+Example: <https://sentry.io/govuk/app-finder-frontend/issues/400074507>
+
+Desired behaviour: backing app returns a 4XX status code, response is fed back to user. Nothing is ever logged or sent to Sentry.
+
+### Manually logged errors
```

Comment:
I think this is really a subset of a broader problem with application logging on GOV.UK.

When the application has got into a state where it has to give up on something, there are things the developer should know. But as a developer debugging a problem, I want to know what the application did *before* it reached that critical point, and even if the application was able to respond I still want to see what it did. Only showing diagnostic information when something has definitely gone wrong makes it harder to debug, because I can't tell if I'm not seeing logs because they didn't get logged, or because I'm using Kibana wrong.

We [deliberately built diagnostic logging into the work we're doing for search indexing](https://github.com/alphagov/rummager/blob/master/lib/govuk_index/publishing_event_worker.rb#L53), because there is a lot of complexity in the system at the moment, and there are a lot of reasons why content can not be updated. Debugging it is hard, but reading these logs is often enough to quickly isolate where the problem is, and verify that yes, this message was received by Rummager and the system isn't magic.

So I agree this is not an appropriate thing to put in sentry and raise alerts for, but I think we could do a lot more to improve application logging to Kibana. I'd like to discuss it more in general.